File: src/Snapik.App/UiSoundService.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Windows.Media;

namespace Snapik.App;

/// <summary>
/// The three interface sounds. <see cref="MediaPlayer"/> instead of <c>SoundPlayer</c>: it has a
/// volume of its own, so the user can turn the sounds down instead of only off, and every sound is
/// mixed at a gain that makes the three sit at the same loudness. The files are ordinary content
/// next to the executable, because MediaPlayer does not open <c>pack://application</c> resources.
/// Every call happens on the UI thread: MediaPlayer belongs to the dispatcher that created it.
/// </summary>
internal static class UiSoundService
{
    private const string ShutterFile = "shutter-1-039s.mp3";
    private const string TickFile = "click-tiny-005s.mp3";
    private const string CopiedFile = "notify-soft-040.mp3";
    // The shutter is mixed well below the other two: it fires on every capture, and the file that
    // replaced the old one is hotter by about 6 dB, so the gain has to give that back and more.
    private static readonly Sound Shutter = new(ShutterFile, 0.6);
    private static readonly Sound TickSound = new(TickFile, 0.25);
    private static readonly Sound CopiedSound = new(CopiedFile, 0.7);
    private static readonly SoundThrottle Throttle = new();

    internal static void Capture(HotkeySettings settings)
    {
        // The shutter mutes the ticks even when the sounds are off: turning them on mid-capture must
        // not let a tick through on the tail of a shutter nobody heard.
        Throttle.Capture(Stopwatch.GetTimestamp());
        if (!settings.PlaySounds) return;
        Shutter.Play(settings.SoundVolume);
    }

    internal static void Tick(HotkeySettings settings)
    {
        if (!settings.PlaySounds) return;
        if (!Throttle.AllowTick(Stopwatch.GetTimestamp())) return;
        TickSound.Play(settings.SoundVolume);
    }

    internal static void Copied(HotkeySettings settings)
    {
        if (!settings.PlaySounds) return;
        CopiedSound.Play(settings.SoundVolume);
    }

    /// <summary>Opens the files while the strip starts, so the first sound is not the one that waits for the disk.</summary>
    internal static void Preload()
    {
        Shutter.Open();
        TickSound.Open();
        CopiedSound.Open();
    }

    /// <summary>Smoke check: the three files are shipped next to the assembly and are really MP3.</summary>
    internal static void VerifyAssets()
    {
        foreach (var fileName in new[] { ShutterFile, TickFile, CopiedFile })
        {
            var path = PathOf(fileName);
            if (!File.Exists(path))
                throw new InvalidOperationException($"The bundled sound \"{fileName}\" is missing next to the assembly.");
            var header = new byte[3];
            using var stream = File.OpenRead(path);
            if (stream.Length == 0 || stream.Read(header, 0, header.Length) != header.Length)
                throw new InvalidOperationException($"The bundled sound \"{fileName}\" is empty.");
            var tagged = header[0] == 0x49 && header[1] == 0x44 && header[2] == 0x33;
            var frameSync = header[0] == 0xFF && (header[1] & 0xE0) == 0xE0;
            if (!tagged && !frameSync)
                throw new InvalidOperationException($"The bundled sound \"{fileName}\" does not start as an MP3 stream.");
        }
    }

    private static string PathOf(string fileName) => Path.Combine(AppContext.BaseDirectory, "Assets", "Audio", fileName);

    private sealed class Sound(string fileName, double gain)
    {
        private MediaPlayer? _player;
        private bool _failed;

        internal void Open()
        {
            // Nothing but the exception of a missing codec or a missing file can happen here, and a
            // sound that cannot be opened must not stop a capture.
            try { _ = Player(); } catch { _failed = true; }
        }

        internal void Play(int volume)
        {
            if (_failed) return;
            try
            {
                var player = Player();
                player.Volume = Math.Clamp(volume, 0, 100) / 100.0 * gain;
                player.Position = TimeSpan.Zero;
                player.Play();
            }
            catch { /* Audio feedback must never interrupt capture or strip interaction. */ }
        }

        private MediaPlayer Player()
        {
            if (_player is { } opened) return opened;
            var player = new MediaPlayer();
            player.MediaFailed += (_, _) => _failed = true;
            // The player is kept before the file is opened, and a throwing Open switches the sound
            // off: otherwise every play would build another MediaPlayer on the same broken file.
            _player = player;
            try { player.Open(new Uri(PathOf(fileName), UriKind.Absolute)); }
            catch { _failed = true; throw; }
            return player;
        }
    }
}

File: macos/Sources/SnapikMac/App/UiSoundService.swift
// Port of `src/Snapik.App/UiSoundService.cs`, SPEC-DELTA-3 §1.5 G-10, G-11. Replaces
// `CaptureFeedbackSound` (two camera WAVs, on or off) with the three sounds of this round, each
// mixed at a gain of its own on top of the `SoundVolume` preference.
import AppKit
import Foundation
import SnapikCore

/// The three interface sounds: the shutter of a capture, the tick of the strip and the note that
/// says the package went to the clipboard. `AVAudioPlayer` and not `NSSound`: it has a volume of its
/// own, so the user can turn the sounds down instead of only off, and the three can be mixed to sit
/// at the same loudness (`UiSoundService.cs`'s `MediaPlayer` for the same reason).
///
/// Called from `App/AppCoordinator+OverlayEditorDelegate.swift`, `App/AppCoordinator+Package.swift`
/// and `Stack/EdgeStackWindowController.swift` (CONTRACTS.md "Shell (звук)").
enum UiSoundService {
    /// The shutter is mixed well below the other two: it fires on every capture, and the file that
    /// replaced the old one is hotter by about 6 dB, so the gain has to give that back and more
    /// (`macos/Resources/AUDIO-README.md` carries the provenance and the measurements).
    private static let shutter = Sound(fileName: "shutter-1-039s", gain: 0.6)
    private static let tickSound = Sound(fileName: "click-tiny-005s", gain: 0.25)
    private static let copiedSound = Sound(fileName: "notify-soft-040", gain: 0.7)

    private static let tickThrottleNanoseconds: UInt64 = 170_000_000
    private static let captureSuppressionNanoseconds: UInt64 = 400_000_000

    private static let gate = NSLock()
    private static var lastCaptureTimestamp: UInt64?
    private static var lastTickTimestamp: UInt64?

    static func capture(_ settings: HotkeySettings) {
        // The shutter mutes the ticks even when the sounds are off: turning them on mid-capture must
        // not let a tick through on the tail of a shutter nobody heard.
        gate.lock()
        lastCaptureTimestamp = DispatchTime.now().uptimeNanoseconds
        gate.unlock()
        guard settings.playSounds else { return }
        shutter.play(volume: settings.soundVolume)
    }

    static func tick(_ settings: HotkeySettings) {
        guard settings.playSounds else { return }
        let now = DispatchTime.now().uptimeNanoseconds
        gate.lock()
        if let captureTimestamp = lastCaptureTimestamp, now - captureTimestamp < captureSuppressionNanoseconds {
            gate.unlock()
            return
        }
        if let previous = lastTickTimestamp, now - previous < tickThrottleNanoseconds {
            gate.unlock()
            return
        }
        lastTickTimestamp = now
        gate.unlock()
        tickSound.play(volume: settings.soundVolume)
    }

    static func copied(_ settings: HotkeySettings) {
        guard settings.playSounds else { return }
        copiedSound.play(volume: settings.soundVolume)
    }

    /// Port of `VerifyAssets`; called first from `SmokeTestRunner` (SPEC §8.4). The three files are
    /// shipped, are not empty and really start as an MP3 stream.
    static func verifyAssets() throws {
        for sound in [shutter, tickSound, copiedSound] {
            guard let data = sound.data(), data.count >= 3 else {
                throw SnapikError.invalidData("The bundled sound \"\(sound.fileName).mp3\" is missing or empty.")
            }
            let header = [UInt8](data.prefix(3))
            let tagged = header[0] == 0x49 && header[1] == 0x44 && header[2] == 0x33
            let frameSync = header[0] == 0xFF && (header[1] & 0xE0) == 0xE0
            guard tagged || frameSync else {
                throw SnapikError.invalidData(
                    "The bundled sound \"\(sound.fileName).mp3\" does not start as an MP3 stream.")
            }
        }
    }

    // MARK: - One sound

    private final class Sound {
        let fileName: String
        private let gain: Double
        private let gate = NSLock()
        private var player: NSSound?
        private var failed = false

        init(fileName: String, gain: Double) {
            self.fileName = fileName
            self.gain = gain
        }

        func data() -> Data? {
            guard let url = Self.resourceURL(named: fileName) else { return nil }
            return try? Data(contentsOf: url)
        }

        /// Port of `Sound.Play`: the volume of the preference, scaled by the gain of this sound, and
        /// the sound restarted from its beginning. Audio feedback must never interrupt a capture, so
        /// a sound that cannot be opened is switched off instead of throwing.
        func play(volume: Int) {
            gate.lock()
            if failed {
                gate.unlock()
                return
            }
            if player == nil {
                guard let data = data(), let opened = NSSound(data: data) else {
                    failed = true
                    gate.unlock()
                    return
                }
                player = opened
            }
            let sound = player
            gate.unlock()

            guard let sound else { return }
            sound.volume = Float(Double(max(0, min(100, volume))) / 100.0 * gain)
            sound.stop()
            sound.play()
        }

        private static func resourceURL(named name: String) -> URL? {
            if let url = Bundle.main.url(forResource: name, withExtension: "mp3") {
                return url
            }
            #if SWIFT_PACKAGE
            if let url = Bundle.module.url(forResource: name, withExtension: "mp3") {
                return url
            }
            #endif
            return nil
        }
    }
}
